packages/fetchai/skills/weather_station/strategy.py

```diff
@@ -92,14 +93,19 @@ def is_matching_supply(self, query: Query) -> bool:
         return True
 
     def generate_proposal_and_data(
-        self, query: Query
+        self, query: Query, counterparty: Address
     ) -> Tuple[Description, Dict[str, List[Dict[str, Any]]]]:
         """
         Generate a proposal matching the query.
 
         :param query: the query
         :return: a tuple of proposal and the weather data
         """
+        tx_nonce = self.context.ledger_apis.generate_tx_nonce(
+            identifier=self._ledger_id,
+            seller=self.context.agent_addresses[self._ledger_id],
+            client=counterparty,
+        )
```

Comment:
We generate the tx_nonce inside the strategy before the proposal.

Reply:
nice!